### DISCUSSION TITLE
Version Upgrade Guide / Software Tool

### DISCUSSION THREAD

**Discussion by gridlocdev** [KinesisCorporation/Adv360-Pro-ZMK#170] [Ideas]
# Version Upgrade Guide / Software Tool
I've come across an identical concern voiced in #167 where new firmware updates regularly include breaking changes that impact user keymap configurations. This requires manual intervention to load the latest updates when updates do occur, and the transition could possibly be a bit easier with better documentation and/or assistive tooling.

## Suggestion

I'd personally like to see one or both of the following:

- Adding version migration helper tools in the web-based GUI keymap editor, such as a button to download an image that includes a visual representation of the full configuration (all layers, keymaps, and macros); which could be used as a reference to help manually re-enter mappings from the existing GUI into a new version of the editor or simply just to share around more easily.
   > _Note: After some brief looking for examples, this is actually an identical idea to the "Print Layout" button of the [ZSA Moonlander editor](https://configure.zsa.io/moonlander/layouts/default/latest/0)_
- Additional help information in the [user manual](https://kinesis-ergo.com/support/kb360pro/) or on the [Firmware Updates section](https://kinesis-ergo.com/support/kb360pro/#firmware-updates) of the Advantage 360 Pro homepage for what individual steps should be taken for a user to upgrade their existing forked firmware to a new version (e.g. v2.0 -> v3.0). From what I can gather, it would involve either of the following two things when a user would like to obtain one of the new version releases:
   1. Performing a "Settings Reset" using the `settings-reset.uf2` file, then manually re-entering all individual keys and macros through the GUI tool
   2. Learning the ZMK keymap language and applying merges through a git merge tool

**Comment by wojo** [comment]
Great ideas, even though I know how to do this manually this would be a major improvement and save me time as well, let alone users that do not have strong knowledge of git, etc 👍

**Comment by vseehausen** [comment]
Yes, please. Untangling firmware updates from users' key customizations should not be a hard technical problem. If you would just prioritize this, it would lead to more happy customers.

**Comment by duanemay** [comment]
After feeling this pain during the upgrade from V1 to V2 and a few other times, I solved the problem for myself. I created a simple file that lists my key customizations and another with my macros. I then made a set of scripts that clones the real firmware repo (a optional param for the branch name), creates the adv360*.keymap files, the keymap.json, and copies the macros to macros,dtsi, it then runs the steps to create the firmware, and copies the generated firmware files to another sub dir, so it doesn't get deleted when you clean up the firmware repo.

If Kinesis provided something like this, I think it would provide a simpler interface for their users. They could just modify the generator scripts when they change formats. The firmware dir, custom mapping file and macro file could all be ignored, so there would be less PRs against the repo with personal changes to the keymap.

**Reply by Omniscently** [reply]
**I will look into having this built.**

Are you willing to share those tools?

I am far from a dev, but if it isn't too difficult to implement I think it would be a valuable addition.

**Reply by duanemay** [reply]
This is the code I have been using...
It is kind of cobbled together in a combination of BASH, Perl, and Golang...
Also, I have not tested it on anything other than a Mac, so will likely be some Mac specific things that need to change.
It works for me, hope this helps others

https://github.com/duanemay/advantageFirmwareGenerator

**Reply by ReFil** [reply]
That's a really cool little program! It's really interesting to see what people are doing with the 360 Pro's firmware
